### ISSUE TITLE
Slow CPU-Intensive Calculations

### ISSUE THREAD

**Issue by CincySteve** [dotnet/runtime#45363]
I have a C# WinForms app that I have been converting to Blazor Wasm. It needs to perform a set of cpu-intensive calculations (i.e. no IO or UI interaction) after most user inputs. The calculations entail repetitively (30-50 times) invoking a number of methods in a set of 25-35 C# class objects depending on the scenario. The same calculation code runs in the WinForms and Blazor apps.

I am seeing a ~20 fold degradation in performance under Blazor (e.g. 350ms in WinForms vs 7000ms in Blazor). That's the case whether run using the debugger or published code.  I have confirmed that the degradation is spread across the calculations versus concentrated in some of the logic.  I gather the degradation might be due to Blazor interpreting versus compiling my C# code, but that the ability to compile may be on the horizon.  So, my questions are:

1.  Does this level of degradation seem right?
2.  Is there a current way to mitigate it?
3.  Assuming the ability to compile is on the way, is there a rough timeframe for that?
4.  If compiling will be a big boost, do you have a rough idea how much slower than a WinForms implementation compiled Blazor Wasm code might be?

I am using Windows 10 on an x64 Dell laptop, VS Community 2019 v16.8.2, AspNetCore 5.0, and Chrome v87.0.4280.66.

**Comment by CincySteve** [comment]
I shifted to discussing this issue in the aspnetcore forum [here](https://github.com/dotnet/aspnetcore/discussions/28080#discussioncomment-258269) and ended up creating a repro demo [here](https://github.com/CincySteve/PerformanceRepro/tree/master).  After some optimization by the person who has been helping me, the repro code runs ~6 times slower in Blazor Wasm versus WinForms.  Part of the optimization (moving away from use of a DataTable) may be less than desirable due to the code changes required.  Anyway, he suggested that I post this back here.

**Comment by CincySteve** [comment]
Can someone let me know what, if anything, is planned for this issue?  How long might it remain "untriaged?"  Thanks.

**Comment by CoffeeFlux** [comment]
cc: @lewing

**Comment by En3Tho** [comment]
@CincySteve What kind of action do you expect? It's known that Blazor is utilizing interpreter now, thus perf degradation is expected. Net6 is going to bring AOT wasm compilation. That will most probably cure perf problems you're experiencing.

**Comment by CincySteve** [comment]
@En3Tho Your response seems to answer my questions.  You believe that my perf difference is due to interpretation.  AOT will "most probably cure perf problems" (without guarantees, of course).  And since AOT is a .Net6 thing, the timing won't be until November, 2021 if I'm recalling correctly.  Unfortunate, but I understand the list of Blazor needs is long.

**Comment by iwis** [comment]
@CincySteve  I found a speed comparison of: Blazor WebAssembly, the current implementation of Blazor WebAssembly AOT, C# running outside the browser, and JavaScript: https://csharp-wasm-benchmark.acmion.com/. Reddit discussion is [here](https://www.reddit.com/r/Blazor/comments/kp1all/c_wasm_benchmark_c_wasm_aot_vs_c_wasm_interpreted/).

I think that it may be helpful in answering some of your questions. I don't program in Blazor, just observe this project, so I don't want to interpret these results.

**Comment by En3Tho** [comment]
@CincySteve Yes You're right. I personally believe Blazor actually needs like 1-2 more years to mature (AOT + more Framework features.. all those routing, generics, nested components improvements etc) Also, WASM too. If calculation time is really noticeable, maybe it's not too unreasonable to actually offload heavy work in JS.

**Comment by KieranDevvs** [comment]
@En3Tho I don't think its necessarily the fact that Blazor is interpreted, is the problem. I read on the AOT issue thread that a lot of optimisations haven't taken place in regards to interpreting thus AOT might not be a holy grail fix for all problems. If I recall correctly, AOT was only going to happen for hot paths of the application and the less critical code can all be interpreted.

>Generally, AoT-compiled code is bigger than interpreted bytecode, with all technologies. This is because interpreters tend to interpret higher-level bytecode than the low-level instructions produced from AoT compilation, so it takes fewer interpreted instructions to represent the same set of acts. You're right that if you went 100% AoT then the interpreter itself could be dropped, but that's not very big compared with the size gains you'll see from AoT-compiling all the framework libraries.
>
>That's not only true in theory, it's also the case in reality :) We see it in practice when AoT-compiling Xamarin apps for iOS/watchOS: the application code becomes quite a bit bigger.
>
>This is why for Blazor WebAssembly, we're investing more effort in creating a mixed-mode AoT system, where applications will only AoT-compile the hottest code paths, leaving the rest as compact but less fast interpreted code. Hopefully we can give developers control over their own size/speed tradeoffs.

**Comment by En3Tho** [comment]
@KieranDevvs I hope we will get an option to choose manually what code gets AOT compiled then. Also, for intranet Enterprise apps I don't think that app size will be a problem. So full AOT is a good choice. So we'll, flexibility is welcome.

**Comment by CincySteve** [comment]
@iwis Thanks for the link to the speed comparison.  Very interesting and comforting, albeit with uncertainty.  @En3Tho I agree that Blazor has a couple of years of maturation to go and agree with your hope to @KieranDevvs that there will be a way for developers to choose what gets compiled versus not given his observations.

**Comment by SamMonoRT** [comment]
Moving this to 7.0.0

**Comment by SamMonoRT** [comment]
@lewing fyi - moved this to 7.0.0

**Comment by CincySteve** [comment]
I have been away from this for quite awhile due to a personal issue and just noticed that this has been moved to 7.0.0.  I'm seeing a projection for 7.0.0 in 11/2022 (not so long from now).  I hope that's the case.  I have just migrated to VS 2022 and .Net 6.  Anyway, still interested in the issue.

**Comment by lewing** [comment]
We've done a considerable amount of work in again net8 to improve computational performance.  If you are interested in trying this again we'd be very interested in hearing your results.

**Comment by CincySteve** [comment]
I just finished trying to shift to net 8 preview, but ran into a VS problem.  I believe I need to upgrade to VS 17.6.5 from 17.5.5. to use net 8.  But 17.6.5 creates [this](https://developercommunity.visualstudio.com/t/Derived-Components-Not-Rendering-in-VS-C/10383214) problem for me and I couldn't get my demo app to run, even though it doesn't use base/derived classes.  I'll try again at some point, but for now I'm not sure how to test the work you have done.  

If you know a way to test net 8 preview using VS 17.5.5, by all means let me how.  Thanks.

P.S.  I did notice a big improvement with net 7 versus 6 (I think I was using 6).
